### PR TITLE
change cache from e2qc to cream

### PR DESCRIPTION
**Pull request by evanmcc**
e2qc had some locking, contention, and speed issues, this is an attempt to swap out the cache library.

### CODE REVIEW

File: rebar.lock [hunk 1]
```diff
@@ -15,15 +15,15 @@
   {git,"https://github.com/helium/clique.git",
        {ref,"e4be0dae150061bec080502ed530091672880867"}},
   0},
+ {<<"cream">>,
+  {git,"https://github.com/JayKickliter/cream",
```
Comment (JayKickliter) [inline rebar.lock:19]: We should probably transfer the repo to the helium org if everything looks good

**Comment by JayKickliter** [comment]
Some context for reviewers: the new caching module is [cream](https://github.com/jaykickliter/cream), which is a NIF wrapper around [Moka](https://github.com/moka-rs/moka).

`e2qc`'s API has some inherent flaws that are not easily worked around: individual caches are referenced by an atom (necessitating a globally locked map holding all caches). Additionally, e2qc caches are lazily constructed when referenced by their atom name, which means that they use default parameters unless you choose provide cache parameters with every single caching call.

File: src/blockchain_sup.erl [hunk 2]
```diff
@@ -143,3 +148,42 @@ init(Args) ->
 %% ------------------------------------------------------------------
 %% Internal Function Definitions
 %% ------------------------------------------------------------------
+
+
+cream_caches_init() ->
+    MaxCapacity = 10000,
+    {ok, VarCache}
+        = cream:new(MaxCapacity,
+                    [{initial_capacity, 1000},
```
Comment (Vagabond) [inline src/blockchain_sup.erl:157]: Are these in byte sizes or in counts?

File: src/blockchain_utils.erl [hunk 3]
```diff
@@ -432,17 +432,7 @@ hex_adjustment(Loc) ->
 -spec score_gateways(Ledger :: blockchain_ledger_v1:ledger()) -> gateway_score_map().
 score_gateways(Ledger) ->
     {ok, Height} = blockchain_ledger_v1:current_height(Ledger),
-    case blockchain_ledger_v1:mode(Ledger) of
```
Comment (Vagabond) [inline src/blockchain_utils.erl:435]: No more cache here?

File: src/blockchain_utils.erl [hunk 4]
```diff
@@ -812,36 +802,40 @@ get_vars(VarList, Ledger) ->
...
+      fun() ->
+              get_var_(VarName, Ledger)
+      end
     ).
 
 -spec get_var(VarName :: atom(), Ledger :: blockchain_ledger_v1:ledger()) -> {ok, any()} | {error, any()}.
 get_var(VarName, Ledger) ->
-    get_var_(VarName, Ledger).
+    VarsNonce =
+        case blockchain_ledger_v1:vars_nonce(Ledger) of
```
Comment (Vagabond) [inline src/blockchain_utils.erl:817]: Did you ever get this to be a optimized lookup value?

File: src/blockchain_sup.erl (hunk 2 above)
Comment (JayKickliter) [reply src/blockchain_sup.erl:157]: item count. cream/main now has the option to bound the cache by memory

File: rebar.config [hunk 5]
```diff
@@ -38,7 +38,7 @@
     {h3, ".*", {git, "https://github.com/helium/erlang-h3.git", {branch, "master"}}},
     {erl_angry_purple_tiger, ".*", {git, "https://github.com/helium/erl_angry_purple_tiger.git", {branch, "master"}}},
     {erlang_stats, ".*", {git, "https://github.com/helium/erlang-stats.git", {branch, "master"}}},
-    {e2qc, ".*", {git, "https://github.com/helium/e2qc", {branch, "master"}}},
+    {cream, ".*", {git, "https://github.com/JayKickliter/cream", {branch, "main"}}},
```
Comment (JayKickliter) [inline rebar.config:41]:
```suggestion
    {cream, ".*", {git, "https://github.com/helium/cream", {branch, "main"}}},
```

**Review by JayKickliter — CHANGES_REQUESTED** [review]
After updating rebar.config to source cream from the helium or, please bump it so that you get the latest API. I don't think you're using anything that changed, but a couple things may break. The biggest change is that you can now bound the cache my [memory] usage.

[memory]: https://github.com/helium/cream/blob/main/README.md?plain=1#L56-L75

File: src/blockchain_sup.erl (hunk 2 above)
Comment (evanmcc) [reply src/blockchain_sup.erl:157]: object count, as I understand it

**Comment by Vagabond** [comment]
I think it would be better to bound by memory and make the memory config tunable from the sys.config

**Comment by JayKickliter** [comment]
> and make the memory config tunable from the sys.config

A global cache size config var or per-cache var?

File: rebar.lock (hunk 1 above)
Comment (JayKickliter) [reply rebar.lock:19]: It's been transferred.

**Comment by Vagabond** [comment]
> > and make the memory config tunable from the sys.config
> 
> A global cache size config var or per-cache var?

I'd prefer a per-cache var

File: src/blockchain_sup.erl [hunk 6]
```diff
@@ -151,21 +151,22 @@ init(Args) ->
 
 
 cream_caches_init() ->
-    MaxCapacity = 10000,
+    MaxItemCapacity = 10000,
+    MB = 1024 * 1024,
     {ok, VarCache}
-        = cream:new(MaxCapacity,
+        = cream:new(MaxItemCapacity,
                     [{initial_capacity, 1000},
```
Comment (JayKickliter) [inline src/blockchain_sup.erl:158]: I added `initial_capacity` for completeness sake, but I'm not sure I'd bother setting it unless you have strong insight into growth. Reallocation is _probably_ fast enough compared to a cache miss.

File: src/grpc/blockchain_grpc_sc_server_handler.erl [hunk 7]
```diff
@@ -115,17 +115,16 @@ maybe_initialize_state(undefined) ->
...
         {ok, 2} ->
             ActiveSCs =
-                e2qc:cache(
-                    ?MODULE,
-                    active_list,
-                   30,
-                    fun() -> maps:to_list(blockchain_state_channels_server:get_actives()) end
-                ),
+                cream:cache(
+                  ?sc_server_cache,
```
Comment (JayKickliter) [inline src/grpc/blockchain_grpc_sc_server_handler.erl:124]: Missing a `persistent_term:get(?sc_server_cache)`?

Comment (JayKickliter) [inline src/grpc/blockchain_grpc_sc_server_handler.erl:124]: Missing a pt:get

```suggestion
                Cache = persistent_term:get(?sc_server_cache),
                cream:cache(
                  Cache,
```

**Review by JayKickliter — CHANGES_REQUESTED** [review]

File: src/blockchain_utils.erl (hunk 4 above)
Comment (evanmcc) [reply src/blockchain_utils.erl:817]: no, I think that this should be a separate PR, as it's a bit tricky to get right.

File: src/blockchain_utils.erl (hunk 3 above)
Comment (evanmcc) [reply src/blockchain_utils.erl:435]: I can't remember my logic here, I'll revert this.

**Review by Vagabond — APPROVED** [review]

**Pull request merged**